File: lang/java/aingle/src/main/java/org/apache/aingle/file/Codec.java
/*

 */
package org.apache.aingle.file;

import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * Interface for AIngle-supported compression codecs for data files.
 *
 * Note that Codec objects may maintain internal state (e.g. buffers) and are
 * not thread safe.
 */
public abstract class Codec {
  /** Name of the codec; written to the file's metadata. */
  public abstract String getName();

  /** Compresses the input data */
  public abstract ByteBuffer compress(ByteBuffer uncompressedData) throws IOException;

  /** Decompress the data */
  public abstract ByteBuffer decompress(ByteBuffer compressedData) throws IOException;

  /**
   * Codecs must implement an equals() method. Two codecs, A and B are equal if:
   * the result of A and B decompressing content compressed by A is the same AND
   * the result of A and B decompressing content compressed by B is the same
   **/
  @Override
  public abstract boolean equals(Object other);

  /**
   * Codecs must implement a hashCode() method that is consistent with equals().
   */
  @Override
  public abstract int hashCode();

  @Override
  public String toString() {
    return getName();
  }

  // Codecs often reference the array inside a ByteBuffer. Compute the offset
  // to the start of data correctly in the case that our ByteBuffer
  // is a slice() of another.
  protected static int computeOffset(ByteBuffer data) {
    return data.arrayOffset() + data.position();
  }
}

File: lang/csharp/src/apache/main/File/Codec.cs
/*

 */

namespace AIngle.File
{
    /// <summary>
    /// Base class for AIngle-supported compression codecs for data files. Note that Codec objects may
    /// maintain internal state (e.g. buffers) and are not thread safe.
    /// </summary>
    public abstract class Codec
    {
        /// <summary>
        /// Compress data using implemented codec
        /// </summary>
        /// <param name="uncompressedData"></param>
        /// <returns></returns>
        abstract public byte[] Compress(byte[] uncompressedData);

        /// <summary>
        /// Decompress data using implemented codec
        /// </summary>
        /// <param name="compressedData"></param>
        /// <returns></returns>
        abstract public byte[] Decompress(byte[] compressedData);

        /// <summary>
        /// Name of this codec type
        /// </summary>
        /// <returns></returns>
        abstract public string GetName();

        /// <summary>
        ///  Codecs must implement an equals() method
        /// </summary>
        /// <param name="other"></param>
        /// <returns></returns>
        abstract public override bool Equals(object other);

        /// <summary>
        /// Codecs must implement a HashCode() method that is
        /// consistent with Equals
        /// </summary>
        /// <returns></returns>
        abstract public override int GetHashCode();

        /// <summary>
        /// Codec types
        /// </summary>
        public enum Type
        {
            /// <summary>
            /// Codec type that implments the "deflate" compression algorithm.
            /// </summary>
            Deflate,

            //Snappy

            /// <summary>
            /// Codec that does not perform any compression.
            /// </summary>
            Null
        };

        /// <summary>
        /// Factory method to return child
        /// codec instance based on Codec.Type
        /// </summary>
        /// <param name="codecType"></param>
        /// <returns></returns>
        public static Codec CreateCodec(Type codecType)
        {
            switch (codecType)
            {
                case Type.Deflate:
                    return new DeflateCodec();
                default:
                    return new NullCodec();
            }
        }

        /// <summary>
        /// Factory method to return child
        /// codec instance based on string type
        /// </summary>
        /// <param name="codecType"></param>
        /// <returns></returns>
        public static Codec CreateCodecFromString(string codecType)
        {
            switch (codecType)
            {
                case DataFileConstants.DeflateCodec:
                    return new DeflateCodec();
                default:
                    return new NullCodec();
            }
        }

        /// <summary>
        /// Returns name of codec
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return GetName();
        }
    }
}
